### DISCUSSION TITLE
How do federated notifications work?

### DISCUSSION THREAD

**Discussion by snarfed** [bluesky-social/atproto#1538] [Protocol (atproto)]
Hi again! Do you all know how cross-PDS notifications will work yet? Eg if I'm on PDS A, and I follow or mention or repost or reply to someone on PDS B, how does PDS B find out?

I looked through all the lexicons, and the federation blog post, but didn't see anything obvious. Seems like something an AppView could do, but I don't know which method it would call on the PDS to deliver notifs. Or do you expect notifications to migrate entirely to AppViews and be served only by them in the future? That would be inconvenient, at least for me, since I'd like notifications to trigger important logic on my PDS.

**Comment by bnewbold** [comment]
At the protocol level, app.bsky notifications are in the AppView, and are still authoritatively pulled via a request (proxied via PDS). The trend has strongly been towards removing application and lexicon awareness from the PDS and in to appview, at least in our reference implementations, which are very focused on big-world.

We recently added native push notifications for mobile platforms, which ping the mobile app and tell it that there may be new notifications to pull. It is possible that the PDS could be involved with push notifications, but they generally require tight integration with the client (app), and the appview is a natural paired service to send the push notifications, for whatever application (potentially multiple separate apps and appviews, each with separate push notification channels, for the same repo/DID).

There is an idea floating around to also have a more generic/standard event stream for things like push notifications, which could work for AppView -> PDS pushes; or PDS -> Client or AppView -> Client.

**Reply by snarfed** [reply]
Hmm, ok. Interesting that the architecture direction is even more toward PDSes being dumb data stores, and all app logic into small-n-federated global services like BGS and AppView. Services like moderation and feeds are notable counterexamples, those are somewhat but not fully app independent, and still separated out into (I'm guessing) a tier you all expect to be maybe medium-n, vs the big-n of PDSes.

I'll keep thinking about where my bridge use case fits. Clearly much of it is the PDS-as-proxy for external users' canonical data in other protocols, but I also need some app-specific logic and realtime (ish, at least) routing/delivery. I'm not sure that fits into any of the obvious categories here yet. The broader, generic event streams you mention definitely sound closest so far!

**Reply by snarfed** [reply]
For now, I guess I can just poll the AppView's `listNotifications` periodically for each of my users. 🤷‍♂️
